### ISSUE TITLE
CLJS instrumentation breaks vararg functions

### ISSUE THREAD

**Issue by ingesolvoll** [metosin/malli#752]
When trying to instrument a CLJS function with a signature like `[x & ys]`, I get an error. 

Given the function

```clojure
(defn my-function-bad
      {:malli/schema [:=> [:cat :int [:* :any]] :any]}
      [x & args]
      (prn "X is " x " args are " args)
      123)
```

Trying to call it with `(my-function-bad 1 2 3)` gives this error:

<img width="496" alt="image" src="https://user-images.githubusercontent.com/426347/190365163-adf0290a-37ef-4b34-97f9-05e8f76e4ede.png">

I made an example project that demonstrates the issue

https://github.com/ingesolvoll/malli-varargs-bug

I suspect this is because we are replacing the original function var using set! in the instrumentation. 1 in this case seems to be bound to the & args of the instrumentation wrapper.

**Comment by dvingo** [comment]
Thanks for the report. I wasn't familiar with the JS output for variadic fns. Here's the compiled code:
```js
malli.instrument_app.my_function_bad = (function malli$instrument_app$my_function_bad(var_args) {
    var args__5775__auto__ = [];
    var len__5769__auto___62323 = arguments.length;
    var i__5770__auto___62324 = (0);
    while (true) {
        if ((i__5770__auto___62324 < len__5769__auto___62323)) {
            args__5775__auto__.push((arguments[i__5770__auto___62324]));

            var G__62325 = (i__5770__auto___62324 + (1));
            i__5770__auto___62324 = G__62325;
            continue;
        } else {
        }
        break;
    }

    var argseq__5776__auto__ = ((((1) < args__5775__auto__.length)) ? (new cljs.core.IndexedSeq(args__5775__auto__.slice((1)), (0), null)) : null);
    return malli.instrument_app.my_function_bad.cljs$core$IFn$_invoke$arity$variadic((arguments[(0)]), argseq__5776__auto__);
});

(malli.instrument_app.my_function_bad.cljs$core$IFn$_invoke$arity$variadic = (function (x, args) {
    cljs.core.prn.cljs$core$IFn$_invoke$arity$variadic(cljs.core.prim_seq.cljs$core$IFn$_invoke$arity$2(["X is ", x, " args are ", args], 0));

    return (123);
}));

(malli.instrument_app.my_function_bad.cljs$lang$maxFixedArity = (1));

/** @this {Function} */
(malli.instrument_app.my_function_bad.cljs$lang$applyTo = (function (seq62276) {
    var G__62277 = cljs.core.first(seq62276);
    var seq62276__$1 = cljs.core.next(seq62276);
    var self__5754__auto__ = this;
    return self__5754__auto__.cljs$core$IFn$_invoke$arity$variadic(G__62277, seq62276__$1);
}));
```
This line is calling the instrumented function with `1` in the first position which is the error we're seeing.
```js
 return malli.instrument_app.my_function_bad.cljs$core$IFn$_invoke$arity$variadic((arguments[(0)]), argseq__5776__auto__);
```

We will need to update the set! call to be aware of arities like spec does:

https://github.com/clojure/clojurescript/blob/a4673b880756531ac5690f7b4721ad76c0810327/src/main/cljs/cljs/spec/test/alpha.cljc#L312

https://github.com/clojure/clojurescript/blob/a4673b880756531ac5690f7b4721ad76c0810327/src/main/cljs/cljs/spec/test/alpha.cljs#L85

**Comment by danielcompton** [comment]
@dvingo, what are your thoughts on wrapping the functions vs annotating them with :pre/:post assertions? I wonder if it would be simpler and have less edge cases?

**Comment by dvingo** [comment]
@danielcompton I think that could get complicated for multi-arity functions where you can have distinct schemas per arity. I just copied what cljs.spec does and it is working well (after figuring out how to correctly pass arguments to the compiled JS code :) ). One nice benefit of this implementation (`set!`'ing each aritiy individually) is that we get support for only instrumenting the arities that the user provided, so you can have some arities where instrumentation doesn't apply.

**Comment by ikitommi** [comment]
Is this fixed via #753?

**Comment by ingesolvoll** [comment]
Yes, at least I can't reproduce this anymore after the fix.
